### PR TITLE
Add splunkhttp.NewHandler

### PR DESCRIPTION
**Pull request by pellared**
## Why 

https://github.com/signalfx/splunk-otel-go/pull/18#issuecomment-808338996

>If we could add some form of this to the distro package. That would seem to make sense to me given it is supposed to be the super mega deluxe easy path for setup Splunk customers.

## What

- Add `splunkhttp.NewHandler`

## Next

- https://github.com/signalfx/splunk-otel-go/pull/30#discussion_r606000990

### CODE REVIEW

File: instrumentation/net/http/splunkhttp/handler.go
Comment (owais) [inline instrumentation/net/http/splunkhttp/handler.go:50]: How about `WithoutServerTiming()` or `WithServerTimingDisabled()`? This wouldn't take any parameters and always set the field to `false`.

File: instrumentation/net/http/splunkhttp/handler.go
Comment (owais) [inline instrumentation/net/http/splunkhttp/handler.go:64]: do we want to keep these options as public considering we have the `With...` helper methods. It's easier to make them public later but impossible to make them private without major version change :)

Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:64]: Struct is not exported so it is not public API

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:50]: It was my initial design. However I feel that current is more flexible if we would like to change the default value or the client would like to get the value from CLI arg using flag package

File: instrumentation/net/http/splunkhttp/handler.go
Comment (owais) [reply instrumentation/net/http/splunkhttp/handler.go:64]: So why not keep fields lower case as well at least for self-documentation?

Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:64]: If we would like to export it at some point then it would be easier 😉 
It is also consistent with:
- https://github.com/open-telemetry/opentelemetry-go-contrib/blob/2760096474280b3634e92431a7ed6da0744ef144/instrumentation/net/http/otelhttp/config.go#L34-L46
- https://github.com/signalfx/splunk-otel-go/blob/59c68c77a30c9c73fff5a8edcb70e4548b7535a2/distro/config.go#L34-L38
It is for me like having `public` methods in `internal` classes in .NET or `package` in Java.

@MrAlias What do you think?

**Comment by pellared** [comment]
@MrAlias Any idea what is going on with MacOS runners? https://github.com/signalfx/splunk-otel-go/pull/30/checks?check_run_id=2246124758 (I cannot investigate it today - I can look tomorrow)

File: instrumentation/net/http/splunkhttp/handler.go
Comment (owais) [reply instrumentation/net/http/splunkhttp/handler.go:64]: We can just update them if we want to expose this later. Imagine another dev in future exposing config struct and then unintentionally exposing all the fields because they were capitalized.

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:30]:
```suggestion
// NewHandler wraps the passed handler in a span named after the operation and with any provided Options.
// This will also enable all the Splunk specific defaults for HTTP tracing.
```

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:43]:
```suggestion
func WithOTelOpts(opts ...otelhttp.Option) Option {
```

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:38]: `operation` is not passed to the `NewHandler`, instead a static `"server"` is used. Is this intentional?

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [reply instrumentation/net/http/splunkhttp/handler.go:64]: It doesn't make too much of a difference at this point as @pellared pointed out.

One nice thing to also mention is that if we continue to unexport `config` but want to unmarshal into it in the future (i.e. yaml or json configuration) these fields will need to be exported.

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:69]:
```suggestion
type optionFunc func(*config)
```

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:78]: `"0"` doesn't seem to be a valid value according to the specification: https://github.com/signalfx/gdi-specification/blob/main/docs/configuration.md#instrumentation-libraries

Should this just check for `false`?

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:63]:
```suggestion
	OTelOpts            []otelhttp.Option
```

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:86]: The precedence of passed configuration overriding environment variables should be tested.

Maybe something similar to [this](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlphttp/options_test.go).

File: instrumentation/net/http/splunkhttp/server_timing_test.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/server_timing_test.go:38]:
```suggestion
func TestNewHandlerDefault(t *testing.T) {
```

File: instrumentation/net/http/splunkhttp/server_timing_test.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/server_timing_test.go:48]:
```suggestion
func TestNewHandlerServerTimingDisabled(t *testing.T) {
```

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/handler.go:43]: This feels like it is going to require a ton of boilerplate to configure things. I wonder if upstream, the `config` should be exported so we could embed it here and just have our options extend upstream. This isn't something blocking this PR, but more so wondering what others thoughts are.

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:26]: I thought it may be good to export. Env vars are kind of public API anyway. Someone may like to use it 😉 I should probably add a comment anyway.

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:38]: It is a 🐛 . Nice catch 👍

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:78]: I believe I saw this pattern of accepting `1` as `true` and `0` as `false` somewhere... But I agree with you. I checked the code and this is e.g how Java parsing is done: via [`Boolean::parseBoolean()`](https://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html)

Will fix that

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:86]: will do

File: instrumentation/net/http/splunkhttp/server_timing_test.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/server_timing_test.go:38]: Will change even though my eyes would even prefer `Test_NewHandler_default` 😄 But I do not have a strong opinion 😉

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:43]: For me it makes sense. I think linters could complain in line :https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/net/http/otelhttp/config.go#L50 as I feel it makes the API less extensible. I can try tackling it after this PR is merged.

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go:69]: I also unexported `apply` in `Option` type to:
```
// Option is used for setting *optional* config properties.
type Option interface {
	apply(*config)
}
```

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Added comments https://github.com/signalfx/splunk-otel-go/pull/30/commits/33dec16e4a16090a3f0b13b567f4f30593df4daf

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Addressed https://github.com/signalfx/splunk-otel-go/pull/30/commits/33dec16e4a16090a3f0b13b567f4f30593df4daf

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Addressed https://github.com/signalfx/splunk-otel-go/pull/30/commits/33dec16e4a16090a3f0b13b567f4f30593df4daf

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Addressed https://github.com/signalfx/splunk-otel-go/pull/30/commits/33dec16e4a16090a3f0b13b567f4f30593df4daf

File: instrumentation/net/http/splunkhttp/server_timing_test.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/server_timing_test.go]: Addressed https://github.com/signalfx/splunk-otel-go/pull/30/commits/33dec16e4a16090a3f0b13b567f4f30593df4daf

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Addressed https://github.com/signalfx/splunk-otel-go/pull/30/commits/33dec16e4a16090a3f0b13b567f4f30593df4daf

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: @MrAlias I just want to let you know that the tests you have send are not really testing the precedence correctly b/c the precedence logic is in test code: https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlphttp/options_test.go#L289-L292

Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Addressed https://github.com/signalfx/splunk-otel-go/pull/30/commits/076f759d6276bbf0b37439611cfcaa420439f180

File: instrumentation/net/http/splunkhttp/server_timing_test.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/server_timing_test.go]: Addressed 33dec16

File: instrumentation/net/http/splunkhttp/handler.go
Comment (MrAlias) [reply instrumentation/net/http/splunkhttp/handler.go]: I'm pretty hesitant to export this at this point. It will set a precedent and I'm not sure it is one we will or won't want in the long term.

I'm not aware of a situation where a user will need to reference this variable given it is expected to describe something set in the runtime environment prior to execution. But, if we do not export this now and there arises a need the future it will be simple enough to export it. In that situation we will probably want to look holistically at exporting all environment variables Splunk uses, likely in some kind of "splunk environment variable" package. If we export this now it will either require us to duplicate the declaration in both packages or be a breaking change.

File: instrumentation/net/http/splunkhttp/config.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/config.go:36]:
```suggestion
// WithServerTiming enables or disables the ServerTimingMiddleware.
//
// The default is to enable the ServerTimingMiddleware if this option is not passed.
// Additionally, the SPLUNK_CONTEXT_SERVER_TIMING_ENABLED environment
// variable can be set to TRUE or FALSE to specify this option. This option value will
// be given precedence if both it and the environment variable are set.
```

File: instrumentation/net/http/splunkhttp/config.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/config.go:43]: Go docs does not interpret this as markdown

```suggestion
// Option is used for setting optional config properties.
```

File: instrumentation/net/http/splunkhttp/config_test.go
Comment (MrAlias) [inline instrumentation/net/http/splunkhttp/config_test.go:31]:
```suggestion
		// Default
```

**Review by MrAlias — COMMENTED** [review]
minor doc edits

File: instrumentation/net/http/splunkhttp/handler.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/handler.go]: Unexported https://github.com/signalfx/splunk-otel-go/pull/30/commits/165c57c14f22bd11afdca17c60347dcc4b9c7208

File: instrumentation/net/http/splunkhttp/config.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/config.go]: Fixed https://github.com/signalfx/splunk-otel-go/pull/30/commits/165c57c14f22bd11afdca17c60347dcc4b9c7208

File: instrumentation/net/http/splunkhttp/config_test.go
Comment (pellared) [reply instrumentation/net/http/splunkhttp/config_test.go]: Fixed https://github.com/signalfx/splunk-otel-go/pull/30/commits/165c57c14f22bd11afdca17c60347dcc4b9c7208

File: instrumentation/net/http/splunkhttp/config.go
Comment (MrAlias) [reply instrumentation/net/http/splunkhttp/config.go]: Did you not want to add the paragraph about defaults and overrides with the environment variable?

Comment (pellared) [reply instrumentation/net/http/splunkhttp/config.go]: You mean https://github.com/signalfx/splunk-otel-go/blob/165c57c14f22bd11afdca17c60347dcc4b9c7208/instrumentation/net/http/splunkhttp/README.md ?

Comment (MrAlias) [reply instrumentation/net/http/splunkhttp/config.go]: Hmm, not sure I follow.

I had originally suggested the `s/(en|dis)abled/\1ables/` change which looks a part of [165c57c](https://github.com/signalfx/splunk-otel-go/commit/165c57c14f22bd11afdca17c60347dcc4b9c7208), but also suggested we add the following as well to the `WithServerTiming` documentation:

```
// The default is to enable the ServerTimingMiddleware if this option is not passed.
// Additionally, the SPLUNK_CONTEXT_SERVER_TIMING_ENABLED environment
// variable can be set to TRUE or FALSE to specify this option. This option value will
// be given precedence if both it and the environment variable are set.
```

Comment (pellared) [reply instrumentation/net/http/splunkhttp/config.go]: It is me how was not following. Give me a minute, please and sorry for inconvenience  😉

Comment (pellared) [reply instrumentation/net/http/splunkhttp/config.go]: Added here: https://github.com/signalfx/splunk-otel-go/pull/30/commits/092496b8b3a7ff804173dc420f8fb27e1b145cb9

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
@owais PTAL

**Review by MrAlias — APPROVED** [review]

**Review by owais — APPROVED** [review]

**Pull request merged**
